master/internal/authz/authz_provider_type.go

```diff
@@ -0,0 +1,63 @@
+package authz
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/determined-ai/determined/master/internal/config"
+	"golang.org/x/exp/maps"
+)
+
+type AuthZProviderType[T any] struct {
+	registry map[string]T
+	once     sync.Once
+}
+
+func (p *AuthZProviderType[T]) Register(authZType string, impl T) {
+	p.once.Do(func() {
+		p.registry = make(map[string]T)
+	})
+	// TODO(ilia): keep registry here or global in internal/authz/authz_basic.go
+	config.RegisterAuthZType(authZType)
+	if _, ok := p.registry[authZType]; ok {
+		panic(fmt.Errorf("can't do double register of type %s for: %s", authZType, p.string()))
+	}
+	p.registry[authZType] = impl
+}
+
+func (p *AuthZProviderType[T]) string() string {
+	return reflect.TypeOf((*T)(nil)).Elem().String()
```

Comment:
nit: Does ```fmt.Sprintf("%T", p.registry["anykey"])``` work here?

I mean they are both pretty gross tbh

Reply:
for non-existing key it prints `<nil>`. we could `fmt.Sprintf("%T", p.registry)` but that'd be prefixed with `map[string]`...